### PR TITLE
chore(transforms): Migrate EventProcessed Events outside of Transforms

### PR DESCRIPTION
**Pull request by Hoverbear**
Builds off #4875 to wrap the event processed event emits out in the topology builder so they don't need to be written per transform. Closes https://github.com/timberio/vector/issues/4931

TODO: Ensure spans are captured right.

### CODE REVIEW

File: src/internal_events/events_processed.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use crate::internal_events::InternalEvent;
+use metrics::counter;
+
+/// A metric to denote the number of events processed by a topology component.
+/// This is wired up already for transforms in the `topology::builder::build_pieces` function,
+/// so you don't need to do that yourself.
+#[derive(Debug)]
+pub struct EventProcessed;
+
+impl InternalEvent for EventProcessed {
```
Comment (lukesteensen) [inline src/internal_events/events_processed.rs:10]: _nit:_ The file name is plural but the struct name is singular :slightly_smiling_face:

File: src/pipeline.rs [hunk 2]
```diff
@@ -17,30 +35,127 @@ impl Sink for Pipeline {
...
+        debug_assert!(self.enqueued.is_empty());
         self.inner.poll_complete()
     }
 }
 
 impl Pipeline {
     #[cfg(test)]
-    pub fn new_test() -> (Self, Receiver<Event>) {
-        Self::new_with_buffer(100)
+    pub fn new_test(inlines: Vec<Box<dyn FunctionTransform>>) -> (Self, Receiver<Event>) {
```
Comment (lukesteensen) [inline src/pipeline.rs:63]: I might prefer we leave this method as-is and add another helper if/when we need it to include inlines. That way we don't need to change everywhere we use this without setting any.

**Review by lukesteensen — COMMENTED** [review]
This is looking pretty good!

File: src/internal_events/events_processed.rs (hunk 1 above)
Comment (Hoverbear) [reply src/internal_events/events_processed.rs:10]: Oh thank you.

File: src/pipeline.rs (hunk 2 above)
Comment (Hoverbear) [reply src/pipeline.rs:63]: This is part of #4875. I'll open a comment linking to this there.

**Comment by Hoverbear** [comment]
image
Seems like the per-transform metrics are working

**Comment by Hoverbear** [comment]
@leebenson I think later the `BytesProcessed` can slot into https://github.com/timberio/vector/blob/a3084ae1359f672b7cae85563a018dcbeb9322d1/src/topology/builder.rs#L125 and https://github.com/timberio/vector/blob/a3084ae1359f672b7cae85563a018dcbeb9322d1/src/topology/builder.rs#L136 ?

**Comment by leebenson** [comment]
👍 Yeah, seems like an ideal place for it.

**Comment by Hoverbear** [comment]
@leebenson Do you want me to throw it in right now?

**Comment by leebenson** [comment]
Please do. Right now, we only get bytes with sinks. Do we need to juice up sources too?

**Comment by Hoverbear** [comment]
Outside the scope of this PR. :) I'll inject a `BytesProcessed` event here.

**Comment by Hoverbear** [comment]
@leebenson How do you want to measure `BytesProcessed` given there is no defined way to measure bytes for an internal structure which may be using referential bytes types or have differing sizes based on encoding?

**Comment by leebenson** [comment]
> @leebenson How do you want to measure BytesProcessed given there is no defined way to measure bytes for an internal structure which may be using referential bytes types or have differing sizes based on encoding?

I assumed just the byte count of the raw, pre-processed incoming payload (like here, measured at the point of `emit!`ing the `EventProcessed`). Is this feasible?

**Comment by Hoverbear** [comment]
Hmmm if we kept our data serialzied in byte buffer or something it would be. But since the values held by `LogEvent` are possibly things like `Bytes` or their own maps, I'm not sure we have an efficient or correct way to say "This event was X number of bytes at the source and is still X number of bytes", or even to say "This even processed Y number of bytes" without having to check the length of every field.

While I think it's possible, I think it's a feature we should consider carefully, as it may require some more architectural work.

**Comment by jszwedko** [comment]
Just dropping a reference to #4705 which discussed similar shortcomings of measuring "bytes processed" when the sink is not doing any serialization. There are some options discussed there.

**Comment by Hoverbear** [comment]
It is, but I'm not really sure how to measure that without actually going and scanning the size of each thing and that being quite expensive?

File: src/topology/builder.rs [hunk 3]
```diff
@@ -130,8 +132,12 @@ pub async fn build_pieces(
             }
             Transform::Task(t) => {
                 let filtered = filter_event_type(input_rx, input_type);
+                let decorated = Box::new(filtered.map(move |event| {
```
Comment (lukesteensen) [inline src/topology/builder.rs:135]: Could be nicer to use `inspect` here rather than `map`.

**Review by lukesteensen — APPROVED** [review]
Looks good to me!

File: src/topology/builder.rs (hunk 3 above)
Comment (Hoverbear) [reply src/topology/builder.rs:135]: Heck, yes!

**Comment by Hoverbear** [comment]
I think this is obsolete.

**Pull request closed without merging**